content/post/2020-07-20-keep.source/index.Rmd

```diff
@@ -0,0 +1,124 @@
+
+Now how do you know where any of your installed packages was installed?
+You can use `find.package()` and `path.package()`!
+
+To check whether a package is installed, it is better to use `find.package()` than `installed.packages()` because the latter, as its docs state, can be slow on some systems.
+In both cases, it does not mean the package is usable, for that you'd need to use `require()`.
+
+## What files are stored locally?
+
+The [R packages book by Hadley Wickham and Jenny Bryan](https://r-pkgs.org/) has [a very neat package called "Package structure and state", including an explanation of the binary state](https://r-pkgs.org/package-structure-state.html#binary-package).
+It says _"There are no .R files in the R/ directory - instead there are three files that store the parsed functions in an efficient file format. This is basically the result of loading all the R code and then saving the functions with save(). (In the process, this adds a little extra metadata to make things as fast as possible)."_
+
+The installed packages in the library are a decompressed version of the binary state.
+There are not all the R files, see [ggplot2 source code](https://github.com/tidyverse/ggplot2/) and ggplot2 on my disk
+
+```{r}
+fs::dir_tree(
+  file.path(
+    find.package("ggplot2"),
+    "R"
+    )
+  )
+```
+
+Under the R folder, there are three files that don't even have the dot R extension!
+
+## How is code stored?
+
+Now, regarding the code, let's mention two important things happening to it.
+
+### Byte compilation
+
+Since [R 3.5](https://cran.r-project.org/doc/manuals/r-release/NEWS.3.html), the code is byte-compiled by default which means it is also stored in a format easier for a machine to deal with.
+You can learn more about byte compilation in the [Efficient R Programming book by Colin Gillespie and Robin Lovelace](https://bookdown.org/csgillespie/efficientR/programming.html#the-byte-compiler), and in [a talk by R Core Member Tomas Kalibera](https://blog.revolutionanalytics.com/2017/08/take-advantage-compiler.html).
+
+### Original formatting and comments?
+
+Then, by default, the source code is stripped of all empty lines and comments because they are useless for code execution and take up space.[^thx]
+
+[^thx]: This came to my attention thanks to [a question by Ofek Shilon on RStudio community](https://community.rstudio.com/t/keep-source-pkgs-vs-keep-source-options/69245).
+
+It is similar to CSS, JS, HTML being minified in web development to make websites load faster.
+Now sometimes you might want to keep code with its comments: 
+as an user for being able to read it locally with all its comments,
+as a developer for debugging or profiling.
+
+As an user installing packages, you need to look into the `keep.source.pkgs` option in `options()` that influences the behavior of package installation, or for a specific package you'd write `install.packages("rhub", INSTALL_opts = "--with-keep.source")`.[^lookup]
```

Comment:
Of course this has no effect when installing binary packages, i.e. for most R users, that get Windows and macOS binaries from CRAN, or Linux binaries from RSPM.